armi/nuclearDataIO/rzflux.py

```diff
@@ -0,0 +1,186 @@
+        self._flux = flux
+        self._metadata = self._getFileMetadata()
+
+    def _getFileMetadata(self):
+        return self._flux.metadata
+
+    @classmethod
+    def _read(cls, fileName: str, fileMode: str) -> RzfluxData:
+        flux = RzfluxData()
+        return cls._readWrite(
+            flux,
+            fileName,
+            fileMode,
+        )
+
+    # pylint: disable=arguments-differ
+    @classmethod
+    def _write(cls, flux: RzfluxData, fileName: str, fileMode: str):
+        return cls._readWrite(flux, fileName, fileMode)
+
+    @classmethod
+    def _readWrite(cls, flux: RzfluxData, fileName: str, fileMode: str) -> RzfluxData:
+        with RzfluxStream(flux, fileName, fileMode) as rw:
+            rw.readWrite()
+        return flux
+
+    def readWrite(self):
+        """
+        Step through the structure of the file and read/write it.
+        """
+        self._rwFileID()
+        self._rw1DRecord()
+        self._rw2DRecord()
+
+    def _rwFileID(self):
+        """
+        Read/write file id record.
+
+        Notes
+        -----
+        The username, version, etc are embedded in this string but it's
+        usually blank. The number 28 was actually obtained from
+        a hex editor and may be code specific.
+        """
+        with self.createRecord() as record:
+            self._metadata["label"] = record.rwString(self._metadata["label"], 28)
+
+    def _rw1DRecord(self):
+        """
+        Read/write File specifications on 1D record.
+
+        This record contains 27 integers.
+        """
+        with self.createRecord() as record:
+            for key in FILE_SPEC_1D_KEYS:
+                # ready for some implicit madness from the FORTRAN 77 days?
+                if key[0] in IMPLICIT_INT:
+                    self._metadata[key] = record.rwInt(self._metadata[key])
+                else:
+                    self._metadata[key] = record.rwFloat(self._metadata[key])
+
+    def _rw2DRecord(self):
+        """
+        Read/write the multigroup fluxes (n/cm^2/2) into a NxG matrix.
+
+        Notes
+        -----
+        Zones are blocked into multiple records so we have to block or unblock
+        them.
+
+        _rwMatrix reverses the indices into FORTRAN data order so be
+        very careful with the indices.
+        """
+        nz = self._metadata["NZONE"]
+        ng = self._metadata["NGROUP"]
+        nb = self._metadata["NBLOK"]
+        if self._flux.groupFluxes is None:
+            # initialize all-zeros here before reading now that we
+            # have the matrix dimension metadata available.
+            self._flux.groupFluxes = numpy.zeros(
+                (ng, nz),
+                dtype=numpy.float32,
+            )
+        blockRatio = (nz - 1) // nb + 1
+        for bi in range(nb):
+            # compute data-blocking bounds, convert to base 0
+            jLow = bi * blockRatio
+            jUp = (bi + 1) * blockRatio
```

Comment:
Should this be `jUp = (bi + 1) * blockRatio - 1`? 

In the original case, if `M = 1` then `JL = 1` and `JUP = blockRatio`. In this case since you are subtracting 1 to start at base 0, would it `bi = 0`, `JL=0`, `JUP = blockRatio - 1`?

Maybe adding a test case where you write back out the binary and compare it to the original will help.

Reply:
Yeah `M = bi + 1` in this implementation, 1 is subtracted from `jLow`, and 1 gets subtracted from the final `jUp`